File: Formularios/Busqueda/frmCheques.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Centrex
{
    public partial class frmCheques
    {
        private ClienteEntity cl = new ClienteEntity();
        private ProveedorEntity pr = new ProveedorEntity();
        private CobroEntity c = new CobroEntity();
        private bool cliente = false;

        public frmCheques(int idCliente = -1, int idProveedor = -1, int idCobro = -1)
        {

            // Esta llamada es exigida por el diseñador.
            InitializeComponent();
            if (idCliente != -1)
            {
                cl = clientes.info_cliente(idCliente);
                cliente = true;
            }
            else
            {
                pr = proveedores.info_proveedor(idProveedor.ToString());
            }

            if (idCobro != -1)
            {
                c = cobros.info_cobro(idCobro.ToString());
            }
            // Agregue cualquier inicialización después de la llamada a InitializeComponent().

        }
        private async void frmCheques_Load(object sender, EventArgs e)
        {
            await InicializarSeleccionChequesAsync();
            await LoadChequesAsync();
        }

        private async void cmd_ok_Click(object sender, EventArgs e)
        {
            if (dg_view.Rows.Count == 0)
                return;

            try
            {
                var ids = dg_view.Rows
                    .Cast<DataGridViewRow>()
                    .Where(fila => fila.Cells["ID"].Value is not null)
                    .Select(fila => Convert.ToInt32(fila.Cells["ID"].Value))
                    .Distinct()
                    .ToList();

                if (ids.Count == 0)
                    return;

                using var ctx = new CentrexDbContext();
                var registros = await ctx.TmpSelChEntity
                    .Where(tmp => ids.Contains(tmp.IdCheque))
                    .ToDictionaryAsync(tmp => tmp.IdCheque);

                foreach (DataGridViewRow fila in dg_view.Rows)
                {
                    if (fila.Cells["ID"].Value is null)
                        continue;

                    var idCheque = Convert.ToInt32(fila.Cells["ID"].Value);
                    var seleccionado = fila.Cells["Seleccionado"].Value != DBNull.Value &&
                                       Convert.ToBoolean(fila.Cells["Seleccionado"].Value);

                    if (registros.TryGetValue(idCheque, out var registro))
                    {
                        registro.Seleccionado = seleccionado;
                    }
                }

                await ctx.SaveChangesAsync();
            }
            catch (Exception ex)
            {
                Interaction.MsgBox($"Error al actualizar la selección de cheques: {ex.Message}", MsgBoxStyle.Critical, "Centrex");
            }
        }

        private async Task InicializarSeleccionChequesAsync()
        {
            if (!cliente)
                return;

            using var ctx = new CentrexDbContext();
            var chequeIds = await ctx.ChequeEntity
                .AsNoTracking()
                .Where(ch => ch.IdCliente == cl.IdCliente)
                .Select(ch => ch.IdCheque)
                .ToListAsync();

            if (chequeIds.Count == 0)
                return;

            var existentes = await ctx.TmpSelChEntity
                .Where(tmp => chequeIds.Contains(tmp.IdCheque))
                .Select(tmp => tmp.IdCheque)
                .ToListAsync();

            var nuevos = chequeIds
                .Except(existentes)
                .Select(id => new TmpSelChEntity
                {
                    IdCheque = id,
                    Seleccionado = false
                })
                .ToList();

            if (nuevos.Count > 0)
            {
                await ctx.TmpSelChEntity.AddRangeAsync(nuevos);
                await ctx.SaveChangesAsync();
            }
        }

        private async Task LoadChequesAsync()
        {
            using var ctx = new CentrexDbContext();

            if (cliente)
            {
                lbl_cheques.Text = "Cheques disponibles del cliente: " + cl.RazonSocial;

                var query = ctx.TmpSelChEntity
                    .AsNoTracking()
                    .Include(tmp => tmp.IdChequeNavigation)
                        .ThenInclude(ch => ch.IdBancoNavigation)
                    .Where(tmp => tmp.IdChequeNavigation.IdCliente == cl.IdCliente)
                    .OrderBy(tmp => tmp.IdCheque)
                    .Select(tmp => new
                    {
                        Seleccionado = tmp.Seleccionado,
                        ID = tmp.IdCheque,
                        BancoEmisor = tmp.IdChequeNavigation.IdBancoNavigation.Nombre,
                        NumeroCheque = tmp.IdChequeNavigation.NCheque,
                        Importe = tmp.IdChequeNavigation.Importe
                    });

                var result = new DataGridQueryResult
                {
                    Query = query,
                    ColumnasOcultar = new List<string> { "ID" }
                };

                await LoadDataGridDynamic.LoadDataGridWithEntityAsync(dg_view, result, depuracion: true);

                if (dg_view.Columns.Contains("BancoEmisor"))
                    dg_view.Columns["BancoEmisor"].HeaderText = "Banco emisor";
                if (dg_view.Columns.Contains("NumeroCheque"))
                    dg_view.Columns["NumeroCheque"].HeaderText = "Nº cheque";
                if (dg_view.Columns.Contains("Importe"))
                {
                    dg_view.Columns["Importe"].HeaderText = "$$";
                    dg_view.Columns["Importe"].DefaultCellStyle.Format = "N2";
                }

                dg_view.ReadOnly = false;
                foreach (DataGridViewColumn columna in dg_view.Columns)
                {
                    columna.ReadOnly = columna.Name != "Seleccionado";
                }
            }
            else
            {
                lbl_cheques.Text = "Cheques en cartera";
                dg_view.DataSource = null;
            }
        }
    }
}

File: __Viejo/Formularios/Busqueda/frmCheques.vb
Public Class frmCheques
    Private cl As New cliente
    Private pr As New proveedor
    Private c As New cobro
    Private cliente As Boolean = False

    Public Sub New(Optional ByVal idCliente As Integer = -1, Optional ByVal idProveedor As Integer = -1, Optional ByVal idCobro As Integer = -1)

        ' Esta llamada es exigida por el diseñador.
        InitializeComponent()
        If idCliente <> -1 Then
            cl = info_cliente(idCliente)
            cliente = True
        Else
            pr = info_proveedor(idProveedor)
        End If

        If idCobro <> -1 Then
            c = info_cobro(idCobro)
        End If
        ' Agregue cualquier inicialización después de la llamada a InitializeComponent().

    End Sub
    Private Sub frmCheques_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        Dim sqlstr As String

        sqlstr = "INSERT INTO tmpSelCH (id_cheque) SELECT id_cheque FROM cheques WHERE id_cliente = '" + cl.id_cliente.ToString + "'"
        ejecutarSQL(sqlstr)

        If cliente Then
            lbl_cheques.Text = "Cheques disponibles del cliente: " + cl.razon_social
            'sqlstr = "SELECT ch.id_cheque AS 'ID', b.nombre AS 'Banco emisor', ch.nCheque AS 'Nº cheque', ch.importe AS '$$' " &
            '            "FROM cheques AS ch " &
            '            "INNER JOIN bancos AS b ON ch.id_banco = b.id_banco " &
            '            "WHERE ch.id_cliente = '" + cl.id_cliente.ToString + "'"
            sqlstr = "SELECT tmpch.seleccionado AS 'Seleccionado', ch.id_cheque AS 'ID', b.nombre AS 'Banco emisor', ch.nCheque AS 'Nº cheque', ch.importe AS '$$' " &
                        "FROM tmpSelCH AS tmpch " &
                        "INNER JOIN cheques AS ch ON tmpch.id_cheque = ch.id_cheque " &
                        "INNER JOIN bancos AS b ON ch.id_banco = b.id_banco " &
                        "WHERE ch.id_cliente = '" + cl.id_cliente.ToString + "' "
            Dim nRegs As Integer = 0
            Dim tPaginas As Integer = 0
            Dim txtnPage As New TextBox()
            cargar_datagrid(dg_view, sqlstr, basedb, 0, nRegs, tPaginas, 1, txtnPage, "cheques", "cheques")

            dg_view.ReadOnly = False
            dg_view.Columns("ID").ReadOnly = True
            dg_view.Columns("Banco emisor").ReadOnly = True
            dg_view.Columns("Nº cheque").ReadOnly = True
            dg_view.Columns("$$").ReadOnly = True
            dg_view.Columns("Seleccionado").ReadOnly = False

        Else
            lbl_cheques.Text = "Cheques en cartera"
        End If
    End Sub

    Private Sub cmd_ok_Click(sender As Object, e As EventArgs) Handles cmd_ok.Click
        Dim sqlstr As String

        If dg_view.Rows.Count > 0 Then
            For Each fila As DataGridViewRow In dg_view.Rows
                If Not fila Is Nothing Then
                    If fila.Cells("Seleccionado").Value Then
                        sqlstr = "UPDATE tmpSelCH SET seleccionado = '1' WHERE id_cheque = '" + fila.Cells("ID").Value.ToString + "'"
                        ejecutarSQL(sqlstr)
                        'MsgBox("Fila " + fila.Cells("Nº cheque").Value.ToString + "seleccionado")
                    End If
                End If
            Next
        End If
    End Sub
End Class